modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/set/optionset/FormOptionSetOptionView.ts

```diff
@@ -202,15 +207,18 @@ module api.form {
                     }
                     this.deselectHandle();
                 }
-                this.notifySelectionChanged();
+                this.notifySelectionChanged(this);
             });
 
             var checkboxEnabledStatusHandler: () => void = () => {
-                var canCheckMoreOptions = !button.isChecked() && this.cantSelectMoreOptions();
-                button.setDisabled(canCheckMoreOptions);
-                button.toggleClass("disabled", canCheckMoreOptions);
+                var buttonShouldBeDisabled = !button.isChecked() && this.cantSelectMoreOptions();
```

Comment:
1. We need to rename the cantSelectMoreOptions method, current name is confusing. Change it to ``isSelectionLimitReached``.

2. Let's add isDisabled method to the Checkbox class:

        isDisabled(): Boolean {
            return this.checkbox.getEl().isDisabled();
        }

3. Add an optional "cls" argument to "setDisabled" method inside the Checkbox class:

        setDisabled(value: boolean, cls?: string): Checkbox {
            this.checkbox.getEl().setDisabled(value);
            if (cls) {
                this.toggleClass(cls, value)
            }
            return this;
        }

4. Extract the code that enables/disables a checkbox into a separate private method (for example "setCheckBoxDisabled") and call it both on initial rendering and inside the checkboxEnabledStatusHandler. 

```Typescript
private setCheckBoxDisabled(checkBox: api.ui.Checkbox) {
  var checkBoxShouldBeDisabled = !checkBox.isChecked() && this.isSelectionLimitReached();

  if (checkBox.isDisabled() != checkBoxShouldBeDisabled) {
    checkBox.setDisabled(checkBoxShouldBeDisabled, "disabled");
  }
}
```